src/AppInstallerCommonCore/Yaml.cpp

```diff
@@ -0,0 +1,478 @@
+    {
+        std::ostringstream out;
+        OutputExceptionHeader(out, type);
+
+        out << (problem ? problem : "Unexplained error");
+
+        m_what = out.str();
+    }
+
+    const char* Exception::what() const noexcept
+    {
+        return m_what.c_str();
+    }
+
+    Node::Node(Type type, std::string tag, const YAML::Mark& mark) :
+        m_type(type), m_tag(std::move(tag)), m_mark(mark)
+    {
+        if (m_type == Type::Sequence)
+        {
+            m_sequence = decltype(m_sequence)::value_type{};
+        }
+        else if (m_type == Type::Mapping)
+        {
+            m_mapping = decltype(m_mapping)::value_type{};
+        }
+    }
+
+    void Node::SetScalar(std::string value)
+    {
+        Require(Type::Scalar);
+        m_scalar = std::move(value);
+    }
+
+    bool Node::operator<(const Node& other) const
+    {
+        Require(Type::Scalar);
+        other.Require(Type::Scalar);
+        return this->m_scalar < other.m_scalar;
+    }
+
+    Node& Node::operator[](std::string_view key)
+    {
+        Require(Type::Mapping);
+        auto itrs = m_mapping->equal_range(key);
+
+        if (itrs.first == itrs.second)
+        {
+            return s_globalInvalidNode;
+        }
+
+        Node& result = itrs.first->second;
+
+        THROW_HR_IF(APPINSTALLER_CLI_ERROR_DUPLICATE_MAPPING_KEY, ++itrs.first != itrs.second);
+
+        return result;
+    }
+
+    const Node& Node::operator[](std::string_view key) const
+    {
+        Require(Type::Mapping);
+        auto itrs = m_mapping->equal_range(key);
+
+        if (itrs.first == itrs.second)
+        {
+            return s_globalInvalidNode;
+        }
+
+        const Node& result = itrs.first->second;
+
+        THROW_HR_IF(APPINSTALLER_CLI_ERROR_DUPLICATE_MAPPING_KEY, ++itrs.first != itrs.second);
+
+        return result;
+    }
+
+    Node& Node::operator[](size_t index)
+    {
+        Require(Type::Sequence);
+        return m_sequence.value()[index];
+    }
+
+    const Node& Node::operator[](size_t index) const
+    {
+        Require(Type::Sequence);
+        return m_sequence.value()[index];
+    }
+
+    size_t Node::size() const
+    {
+        switch (m_type)
+        {
+        case Type::Invalid:
+        case Type::None:
+        case Type::Scalar:
+            return 0;
+        case Type::Sequence:
+            return m_sequence->size();
+        case Type::Mapping:
+            return m_mapping->size();
+        }
+
+        THROW_HR(E_UNEXPECTED);
+    }
+
+    const std::vector<Node>& Node::Sequence() const
+    {
+        Require(Type::Sequence);
+        return m_sequence.value();
+    }
+
+    const std::multimap<Node, Node>& Node::Mapping() const
+    {
+        Require(Type::Mapping);
+        return m_mapping.value();
+    }
+
+    void Node::Require(Type type) const
+    {
+        THROW_HR_IF(APPINSTALLER_CLI_ERROR_YAML_INVALID_OPERATION, m_type != type);
+    }
+
+    void Node::Require(Type type1, Type type2) const

```

Comment:
>Require [](start = 15, length = 7)

I think this is never used or not even possible?

Reply:
It was used previously, but I had to change the code.  It is possible, it is simply !(type1 || type2).

---
In reply to: [468995615](https://github.com/Microsoft/pkg/pull/535#discussion_r468995615) [](ancestors = 468995615)